### PR TITLE
allow for control being removed from visual tree in :pointerover logic

### LINKED ISSUE #1997: changing ListBoxItem template on :pointerover is not always reset when mouse leaves
The following code can leave some lines in the mouseover state when the mouse is moved quickly

```
<ListBox >
        <ListBox.Items xmlns:sys="clr-namespace:System;assembly=mscorlib">
          <sys:String>Hello</sys:String>
          <sys:String>Cruel</sys:String>
          <sys:String>World</sys:String>
          <sys:String>Hello</sys:String>
          <sys:String>Cruel</sys:String>
          <sys:String>World</sys:String>
          <sys:String>Hello</sys:String>
          <sys:String>Cruel</sys:String>
          <sys:String>World</sys:String>
          <sys:String>Hello</sys:String>
          <sys:String>Cruel</sys:String>
          <sys:String>World</sys:String>
          <sys:String>Hello</sys:String>
          <sys:String>Cruel</sys:String>
          <sys:String>World</sys:String>
          <sys:String>Hello</sys:String>
          <sys:String>Cruel</sys:String>
          <sys:String>World</sys:String>
          <sys:String>Hello</sys:String>
          <sys:String>Cruel</sys:String>
          <sys:String>World</sys:String>
        </ListBox.Items>
        
        <ListBox.Styles>
          <Style Selector="ListBoxItem:pointerover /template/ ContentPresenter">
            <Setter Property="Background" Value="Green"/>
            <Setter Property="ContentTemplate" >
              <DataTemplate>
                <StackPanel Orientation="Horizontal" Height="24">
                    <TextBlock Width="80" Text="{Binding}"></TextBlock>
                  <Button >
                    <TextBlock Text="->"/>
                  </Button>
                  <Button >
                    <TextBlock Text="+" />
                  </Button>
                </StackPanel>
              </DataTemplate>
            </Setter>
          </Style>
          
        </ListBox.Styles>
        <ListBox.ItemsPanel>
          <ItemsPanelTemplate>
            <StackPanel Spacing="2" />
          </ItemsPanelTemplate>
        </ListBox.ItemsPanel>
        <ListBox.ItemTemplate >
          <DataTemplate >
            <StackPanel Orientation="Horizontal">
              <StackPanel Orientation="Horizontal" Height="24">
                <TextBlock Width="80" Text="{Binding}"></TextBlock>
              </StackPanel>
            </StackPanel>
          </DataTemplate>
        </ListBox.ItemTemplate>
      </ListBox>
```

### LINKED ISSUE #2079: :pointerover logic does not respond to content changing or moving without mouse movement
Currently IsPointerOver and :pointerover are only updated on mouse movement. Changing content under a stationary mouse has no effect. wpf does respond to content changing.

### PR DESCRIPTION
**Pull request by ahopper**
- What does the pull request do?
Fixes issues with pointerover status
- What is the current behavior?
IsPointerOver & :pointerover can be left wrongly set if the visual tree is modified between mouse movements 
- What is the updated/expected behavior with this PR?
:pointerover is correct
- How was the solution implemented (if it's not obvious)?
in the case of the saved pointerover element being removed from the visualtree a slower top down search is used to tidy the tree of IsPointerOver values
Checklist:

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

If the pull request fixes issue(s) list them like this:

Fixes #1997 
Fixes #2079

### CODE REVIEW

**Comment by ahopper** [comment]
This fix does not try to fix #2079 and it is probable that a fix for that would be a neater solution and render some or all of this redundant.  I'm not sure of the best way to coordinate layout changes with MouseDevice

**Comment by ahopper** [comment]
I now think the changes here won't change in a fix for #2079

**Review by grokys — CHANGES_REQUESTED** [review]
Just tried this and it didn't seem to work. Repo:

- Open Button page in control catalog
- Open devtools and show button in tree by pressing ctrl+shift
- Make sure sidebar is selected
- Hover over button: notice that the button has `:pointerover`
- Use the arrow keys to navigate to a different page
- The button still has `:pointerover`

**Comment by ahopper** [comment]
yep it does not currently try to fixup items removed from the visual tree, it just prevents anything remaining visual not being cleared.  I'm working on the second stage of this that copes with content changing under the mouse and was going to put that in a different pr. That might be the place to look at this tidy up.

**Comment by grokys** [comment]
> it just prevents anything remaining visual not being cleared

Could you explain this? Not sure I understand.

**Comment by ahopper** [comment]
this only tries to fix the visible problem in #1997 plus the more serious hit test issue and I think it does that, previously visible items could be left with :pointerover set wrongly. In fixing it I realized the other issue of content changing without mouse movement the fix for which depends on this fix and would clean up items removed from the visual tree. If you like we can treat this as wip and I could add the other fixes here but then it might be worth doing #2090 as a priority as it is a serious issue.

**Comment by ahopper** [comment]
btw I found yet another :pointerover issue today when playing with mouse wheel scrolling a virtualized  list box, the pointerover effect scrolls away from the pointer and can popup again in a recycled container at the other end of the list when it reaches the top/bottom.

**Comment by grokys** [comment]
Ok, I'm going to merge #2090 while we're talking about/iterating on this as that PR is simpler.

**Comment by ahopper** [comment]
This now fixes #2079 and clears pointerover from items removed from the visual tree, I am triggering the update in TopLevel Invalidate, I'm not sure this is the best place.  There are some side effect, in the render sample the behavior of the bouncing changes as it bounces out of the pointer over condition and resets.

File: src/Avalonia.Input/IPointerDevice.cs [hunk 1]
```diff
@@ -12,5 +12,7 @@ public interface IPointerDevice : IInputDevice
         void Capture(IInputElement control);
 
         Point GetPosition(IVisual relativeTo);
+
+        void LayoutUpdated(IInputRoot root);
```
Comment (grokys) [inline src/Avalonia.Input/IPointerDevice.cs]: I don't think `LayoutUpdated` is a good name for this: it suggests a change in the layout of the scene, but this isn't actually related to layout. Something like `SceneInvalidated` would be a better name I think.

Also I'm not really sure about putting this in the `IPointerDevice` interface. That interface is user-facing but this represents an implementation detail. Not really sure what a better approach would be. @jkoritzinsky what do you think?

File: src/Avalonia.Input/MouseDevice.cs [hunk 2]
```diff
@@ -104,6 +104,18 @@ public void ProcessRawEvent(RawInputEventArgs e)
                 ProcessRawEvent(margs);
         }
 
+        public void LayoutUpdated(IInputRoot root)
+        {
+            if (Captured == null)
```
Comment (grokys) [inline src/Avalonia.Input/MouseDevice.cs]: It feels like we can be a bit more intelligent here. Currently, if I understand correctly we're recalculating the pointer-over state on every invalidation of the scene: that's going to happen _really_ often. We should be able to tell if the scene invalidation affects the point where the mouse currently is, right?

**Review by grokys — CHANGES_REQUESTED** [review]
Thanks for this! It's getting there, but given that this is pretty core behavior that will be called very often I think we need to think a little more about it.

File: src/Avalonia.Input/MouseDevice.cs (hunk 2 above)
Comment (ahopper) [reply src/Avalonia.Input/MouseDevice.cs]: that would be great, but is not an area of avalonia I have much of a feel for yet.

Comment (ahopper) [reply src/Avalonia.Input/MouseDevice.cs]: we could pass the invalidated rect and compare to the last mouse pos, there is also the issue of the window moving to think of.

Comment (grokys) [reply src/Avalonia.Input/MouseDevice.cs]: Yeah that's along the lines of what I was thinking. Isn't the pointer position already in windows coordinates, so moving windows shouldn't have an effect?

Comment (ahopper) [reply src/Avalonia.Input/MouseDevice.cs]: if the window moves by some means other than mouse interaction , the mouse will effectively move relative to the window without mousemove firing and I'm not sure if the scene is invalidated. Not a common thing :)

**Comment by grokys** [comment]
Sorry for the delay in getting back to this @ahopper. I just tried the repro from my first review and it still doesn't seem to work. What am I missing?

**Comment by ahopper** [comment]
@grokys you aren't missing anything, I appear to have broken it with one of the last mods, I did have it fixing you repro previously. I'll try and find some time in the next couple of days to sort it.

**Comment by ahopper** [comment]
Toplevel IRenderRoot.Invalidate(Rect rect) does not seem to be called anymore which explains why this pr is broken

**Comment by grokys** [comment]
@ahopper I've added an event which is called when the scene is invalidated [as discussed by @kekekeks in gitter](https://gitter.im/AvaloniaUI/Avalonia?at=5c1622b3392d8c6f8e6cb9d6). Could you take a look to see if it makes sense?

**Comment by ahopper** [comment]
@grokys that makes sense to me.

**Review by grokys — APPROVED** [review]

**Comment by grokys** [comment]
Thanks @ahopper - merged!

**Pull request merged**